bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/SortedLedgerStorageCheckpointTest.java

```diff
@@ -0,0 +1,231 @@
+            throws IOException {
+            log.info("Complete checkpoint : {}", checkpoint);
+        }
+    }
+
+    private SortedLedgerStorage storage;
+    private Checkpointer checkpointer;
+    private final LinkedBlockingQueue<Checkpoint> checkpoints;
+    private final TestCheckpointSource checkpointSrc = new TestCheckpointSource();
+
+    public SortedLedgerStorageCheckpointTest() {
+        super();
+        conf.setEntryLogSizeLimit(1);
+        this.checkpoints = new LinkedBlockingQueue<>();
+    }
+
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+
+        // initial checkpoint
+
+        this.storage = new SortedLedgerStorage();
+        this.checkpointer = checkpoint -> storage.getScheduler().submit(() -> {
+            log.info("Checkpoint the storage at {}", checkpoint);
+            try {
+                storage.checkpoint(checkpoint);
+                checkpoints.add(checkpoint);
+            } catch (IOException e) {
+                log.error("Failed to checkpoint at {}", checkpoint, e);
+            }
+        });
+        this.storage.initialize(
+            conf,
+            mock(LedgerManager.class),
+            ledgerDirsManager,
+            ledgerDirsManager,
+            checkpointSrc,
+            checkpointer,
+            NullStatsLogger.INSTANCE);
+    }
+
+    @After
+    @Override
+    public void tearDown() throws Exception {
+        if (null != storage) {
+            storage.shutdown();
+        }
+        super.tearDown();
+    }
+
+    ByteBuf prepareEntry(long ledgerId, long entryId) {
+        ByteBuf entry = Unpooled.buffer(4 * Long.BYTES);
+        // ledger id, entry id, lac
+        entry.writeLong(ledgerId);
+        entry.writeLong(entryId);
+        entry.writeLong(entryId - 1);
+        // data
+        entry.writeLong(entryId);
+        return entry;
+    }
+
+    @Test
+    public void testCheckpoint() throws Exception {
+        // memory table holds the first checkpoint, but it is not completed yet.
+        Checkpoint memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+
+        // write entries into ledger storage
+        long lid = System.currentTimeMillis();
+        storage.setMasterKey(lid, new byte[0]);
+        for (int i = 0; i < 20; i++) {
+            storage.addEntry(prepareEntry(lid, i));
+        }
+        // simulate journal persists the entries in journal;
+        checkpointSrc.advanceOffset(100);
+
+        // memory table holds the first checkpoint, but it is not completed yet.
+        memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+
+        // trigger a memtable flush
+        storage.onSizeLimitReached(checkpointSrc.newCheckpoint());
+        // wait for checkpoint to complete
+        checkpoints.poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
```

Comment:
timeout values here are meaningless

Reply:
yes. the test case just waits for checkpoint to complete.